src/java/org/apache/cassandra/dht/tokenallocator/IsolatedTokenAllocator.java

```diff
@@ -0,0 +1,156 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Maps;
+import org.apache.commons.math3.stat.descriptive.SummaryStatistics;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.locator.AbstractReplicationStrategy;
+import org.apache.cassandra.locator.InetAddressAndPort;
+import org.apache.cassandra.locator.SimpleSnitch;
+import org.apache.cassandra.locator.TokenMetadata;
+
+/**
+ * A utility class for allocating a set number of new tokens given an existing {@link TokenMetadata} object without
+ * changing the source metadata.
+ */
+public class IsolatedTokenAllocator
+{
+    private static final Logger logger = LoggerFactory.getLogger(IsolatedTokenAllocator.class);
+
+    public static Collection<Token> allocateTokens(int additionalSplits, AbstractReplicationStrategy source)
+    {
+        Preconditions.checkArgument(additionalSplits > 0, "additionalSplits must be greater than zero");
+        Preconditions.checkNotNull(source);
+
+        List<Token> allocatedTokens = new ArrayList<>();
+        QuietAllocator allocator = new QuietAllocator(source);
+
+        // Distribute nodes among the racks in round-robin fashion in the order the user is supposed to start them.
+        var localDc = source.snitch.getLocalDatacenter();
+        // Get a list to consistently iterate over the racks as we allocate nodes
+        List<String> racks = source.getTokenMetadata().getTopology().getDatacenterRacks().get(localDc).keys().elementSet().asList();
+        int nodeId = 0;
+        int tokensAdded = 0;
+        while (tokensAdded <= additionalSplits)
+        {
+            // Allocate tokens for current node, distributing tokens round-robin among the racks.
+            var newTokens = allocator.allocateTokensForNode(nodeId, racks.get(nodeId % racks.size()));
+            tokensAdded += newTokens.size();
+            allocatedTokens.addAll(newTokens);
+            nodeId++;
+        }
+        return allocatedTokens;
+    }
+
+    /**
+     * A token allocator that takes a source token metadata and replication strategy, but clones the source token
+     * metadata with a quiet snitch, so that added nodes are not communicated to the rest of the system.
+     */
+    private static class QuietAllocator
+    {
+        private final QuietSnitch quietSnitch;
+        private final TokenMetadata quietTokenMetadata;
+        private final TokenAllocation allocation;
+        private final Map<String, SummaryStatistics> lastCheckPoint = Maps.newHashMap();
+
+        private QuietAllocator(AbstractReplicationStrategy source)
+        {
+            this.quietSnitch = new QuietSnitch();
+            this.quietTokenMetadata = source.getTokenMetadata().cloneWithNewSnitch(quietSnitch);
+            // TODO all or full replicas?
```

Comment:
Full. AFAIU transient replicas should not affect the allocation of tokens for normal ones, which we want to properly distribute for.

(This does not matter at all for CNDB.)